elements/rh-menu/rh-menu.ts

```diff
@@ -0,0 +1,267 @@
+  @property({ type: Boolean }) disabled = false;
+
+  /** keeps menu open after a menu item has been clicked  */
+  @property({ type: Boolean, attribute: 'keep-open-on-click' })
+    keepOpenOnClick = false;
+
+  /** position of menu, relative to invoking button */
+  @property() position: Placement = 'bottom';
+
+  @colorContextProvider()
+  @property({ reflect: true, attribute: 'color-palette' }) colorPalette?: ColorPalette;
+
+  @colorContextConsumer() private on?: ColorTheme;
+
+  #tabindex = new RovingTabindexController(this);
+
+  /** if mouse is hovering on menu or button  */
+  #hover = false;
+
+  /** if focus is on button or a menu items  */
+  #focus = false;
+
+  /** if menumitems have been initialized  */
+  #init = false;
+
+  /** menu button element  */
+  #menuButton?: HTMLElement;
+
+  #lastActive?:HTMLElement;
+
+  #float = new FloatingDOMController(this, {
+    content: (): HTMLElement | undefined | null => this.shadowRoot?.querySelector('#button')
+  });
+
+  get open() {
+    const { open } = this.#float;
+    return !!open;
+  }
+
+  firstUpdated() {
+    this.#initMenuButton();
+    if (!this.#init) {
+      this.#initMenuItems();
+    }
+  }
+
+  connectedCallback() {
+    super.connectedCallback();
+    this.addEventListener('click', this.#onClick);
+    this.addEventListener('keydown', this.#onKeydown);
+  }
+
+  render() {
+    const { on = '', hidden, disabled } = this;
+    const { alignment, anchor, open, styles } = this.#float;
+    return html`
+      <div id="container"
+           style="${styleMap(styles)}"
+           class="${classMap({
+             open,
+             [on]: !!on,
+             [anchor]: !!anchor,
+             [alignment]: !!alignment })}">
+        <slot id="button" name="button"></slot>
+        <div part="menu"
+             role="menu"
+             aria-labelledby="button"
+             aria-hidden="${String(!open) as 'true'|'false'}"
+             ?disabled=${!!disabled || !open}
+             ?hidden="${!!hidden}"
+             @focusin=${this.#onFocusin}
+             @focusout=${this.#onFocusout}
+             @mouseover=${this.#onMouseover}
+             @mouseout=${this.#onMouseout}>
+          <slot></slot>
+        </div>
+      </div>
+    `;
+  }
+
+  updated(changedProperties:Map<string, unknown>) {
+    if (changedProperties.has('disabled')) {
+      this.#menuButton?.toggleAttribute('disabled', this.disabled);
+    }
+    if (changedProperties.has('hidden')) {
+      this.#menuButton?.toggleAttribute('hidden', this.hidden);
+    }
+  }
+
+  /**
+   * given button slot, finds menu button and sets attributes accordingly
+   */
+  #initMenuButton() {
+    this.#menuButton = this.#getSlottedButton(this.querySelector('[slot="button"]')) ?? undefined;
+    if (this.#menuButton) {
+      this.disabled = this.#menuButton.hasAttribute('disabled');
+      this.hidden = this.#menuButton.hasAttribute('hidden');
+      this.#menuButton.setAttribute('aria-controls', 'menu');
```

Comment:
we can't do cross root aria like this, so we'll need a workaround.

Reply:
maybe we can put the aria attrs on the wrapping elements in shadow dom, but i'm not sure if that's legit.

we can use ARIAMixin IDL reflection but that's not supported in FF, and i'm not sure if it's polyfillable.

Reply:
we might have better luck by separating the listbox / tabindex part from the button part, then use ElementInternals / ARIA to connect both parts via `for` attr or something like that

Reply:
ok [8c166b8](https://github.com/RedHat-UX/red-hat-design-system/pull/679/commits/8c166b816a9c6e2c789e9747b99353ced77e425f) attempts to solve this by giving the rh-menu host the menu role, and autogenerating IDs on the toggle and on the host. It also adds a `toggle` attr which is a DOMString id to the in-same-root toggle button

```html
<rh-button id="toggle-it">Menu!</rh-button>
<img src="tada.gif" role="presentation">
<rh-menu toggle="toggle-it">...</rh-menu>
```

this doesn't deal with activedescendant though.

Reply:
I've got a commit locally which removes the button and popup from menu and moves that stuff into audio-player. I think this will be the most straightforward path to a11y at the moment. as of this comment there's a lingering issue with the getting FloatingDOMController to render things exactly as their appear in the DP so maybe we can pair of that

Reply:
I was able to make some adjustments to what you had to get the floatingDOMController working, but, as always, I am happy to get any feedback.